### PR TITLE
KAFKA-16649: Remove lock from DynamicBrokerConfig.removeReconfigurable

### PR DESCRIPTION
**Pull request by cmccabe**
Do not acquire the DynamicBrokerConfig lock in DynamicBrokerConfig.removeReconfigurable. It's not necessary, because the list that these functions are modifying is a thread-safe CopyOnWriteArrayList.  In DynamicBrokerConfig.reloadUpdatedFilesWithoutConfigChange, I changed the code to use a simple Java forEach rather than a Scala conversion, in order to feel more confident that concurrent modifications to the List would not have any bad effects here. (forEach is always safe on CopyOnWriteArrayList.)

### CODE REVIEW

File: core/src/main/scala/kafka/server/DynamicBrokerConfig.scala [hunk 1]
```diff
@@ -303,17 +303,17 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
     addBrokerReconfigurable(controller.socketServer)
   }
 
-  def addReconfigurable(reconfigurable: Reconfigurable): Unit = CoreUtils.inWriteLock(lock) {
+  def addReconfigurable(reconfigurable: Reconfigurable): Unit = {
     verifyReconfigurableConfigs(reconfigurable.reconfigurableConfigs.asScala)
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:298]: It seems `verifyReconfigurableConfigs` does not need lock also, since it just check the input strings.

File: core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala [hunk 2]
```diff
@@ -1541,6 +1541,36 @@ class KRaftClusterTest {
...
+      cluster.waitForReadyBrokers()
+      val admin = Admin.create(cluster.clientProperties())
+      try {
+        admin.incrementalAlterConfigs(
+          Collections.singletonMap(new ConfigResource(Type.BROKER, ""),
+            Collections.singletonList(new AlterConfigOp(
+              new ConfigEntry(KafkaConfig.NumNetworkThreadsProp, "2"), OpType.SET)))).all().get()
+        val newTopic = Collections.singletonList(new NewTopic("test-topic", 1, 1.toShort))
+        val createTopicResult = admin.createTopics(newTopic)
+        createTopicResult.all().get()
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:1647]: It seems reducing the network threads causes the disconnection and it produces the retry of request forward. Hence, that can make this test unstable. Maybe we can increase the network threads instead?

File: core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala [hunk 3]
```diff
@@ -1541,6 +1541,36 @@ class KRaftClusterTest {
       cluster.close()
     }
   }
+
+  @Test
+  def testReduceNumNetworkThreads(): Unit = {
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala]: Pardon me, is this test case related to deadlock?

**Review by chia7712 — COMMENTED** [review]
@cmccabe nice improvement. some trivial comments left.

**Review by mumrah — COMMENTED** [review]
Thanks for the patch, @cmccabe. I looked for other usages of this lock and see we're obtaining the write lock in `DynamicBrokerConfig#updateBrokerConfig` which gets called from DynamicConfigPublisher. Will we still need this locking when we have dropped ZK? 

Can you add a comment next to the `lock` indicating what it is guarding?

File: core/src/main/scala/kafka/server/DynamicBrokerConfig.scala (hunk 1 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:298]: Good catch.

File: core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala (hunk 3 above)
Comment (cmccabe) [reply core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala]: I wanted to ensure that no deadlocks are created in the future, by adding this test coverage. I don't think this PR is required to avoid deadlock currently, however.

**Comment by cmccabe** [comment]
> Thanks for the patch, @cmccabe. I looked for other usages of this lock and see we're obtaining the write lock in DynamicBrokerConfig#updateBrokerConfig which gets called from DynamicConfigPublisher. Will we still need this locking when we have dropped ZK?

I think we can simplify this code a lot once ZK is gone, yes. Having a single updater will help a lot. I suspect we can get rid of the lock at that point, although I'll have to look more later.

> Can you add a comment next to the lock indicating what it is guarding?

I added a clarification that the lock protects against concurrent reconfiguration operations, but not concurrent additions or removals of reconfigurables. As you mentioned, this is mostly only relevant to ZK mode since there we can have racing threads.

**Comment by mumrah** [comment]
The new test seems to be failing: [`testReduceNumNetworkThreads() – kafka.server.KRaftClusterTest`](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-15838/1/tests) can you check on it?

**Comment by chia7712** [comment]
> The new test seems to be failing: [testReduceNumNetworkThreads() – kafka.server.KRaftClusterTest](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-15838/1/tests) can you check on it?

see https://github.com/apache/kafka/pull/15838#discussion_r1585911858

File: core/src/main/scala/kafka/server/DynamicBrokerConfig.scala (hunk 1 above)
Comment (chia7712) [reply core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:298]: @cmccabe Can we remove the lock from `verifyReconfigurableConfigs` in this PR? Or we can address that in next one.

**Comment by cmccabe** [comment]
The test should be fixed. I rebased on trunk. Thanks for the reviews @mumrah and @chia7712

File: core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala [hunk 4]
```diff
@@ -1579,6 +1579,36 @@ class KRaftClusterTest {
...
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala]: Could you please rename it to `testIncreaseNumNetworkThreads` by virtue of recent change?

**Review by chia7712 — APPROVED** [review]
@cmccabe thanks for updated PR. please take a look at one small comment

File: core/src/main/scala/kafka/server/DynamicBrokerConfig.scala (hunk 1 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:298]: done

**Comment by chia7712** [comment]
@cmccabe could you please rebase code to include the fix #16857?

**Comment by cmccabe** [comment]
> @cmccabe could you please rebase code to include the fix https://github.com/apache/kafka/pull/16857?

rebased

**Review by omkreddy — APPROVED** [review]
